File: myExpenses/src/main/java/org/totschnig/myexpenses/fragment/SplitPartList.java
/*   This file is part of My Expenses.
 *   My Expenses is free software: you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License as published by
 *   the Free Software Foundation, either version 3 of the License, or
 *   (at your option) any later version.
 *
 *   My Expenses is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License
 *   along with My Expenses.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.totschnig.myexpenses.fragment;

import android.content.Intent;
import android.database.Cursor;
import android.net.Uri;
import android.os.Bundle;
import android.view.ContextMenu;
import android.view.ContextMenu.ContextMenuInfo;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView.AdapterContextMenuInfo;

import org.totschnig.myexpenses.MyApplication;
import org.totschnig.myexpenses.R;
import org.totschnig.myexpenses.activity.ExpenseEdit;
import org.totschnig.myexpenses.activity.ProtectedFragmentActivity;
import org.totschnig.myexpenses.adapter.SplitPartAdapter;
import org.totschnig.myexpenses.databinding.SplitPartsListBinding;
import org.totschnig.myexpenses.model.Money;
import org.totschnig.myexpenses.provider.TransactionProvider;
import org.totschnig.myexpenses.task.TaskExecutionFragment;
import org.totschnig.myexpenses.util.CurrencyFormatter;
import org.totschnig.myexpenses.util.TextUtils;
import org.totschnig.myexpenses.util.Utils;
import org.totschnig.myexpenses.viewmodel.data.Account;

import javax.inject.Inject;

import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import androidx.loader.app.LoaderManager;
import androidx.loader.content.CursorLoader;
import androidx.loader.content.Loader;
import icepick.Icepick;
import icepick.State;

import static org.totschnig.myexpenses.activity.ConstantsKt.EDIT_REQUEST;
import static org.totschnig.myexpenses.provider.DatabaseConstants.KEY_ACCOUNTID;
import static org.totschnig.myexpenses.provider.DatabaseConstants.KEY_AMOUNT;
import static org.totschnig.myexpenses.provider.DatabaseConstants.KEY_LABEL_MAIN;
import static org.totschnig.myexpenses.provider.DatabaseConstants.KEY_PARENTID;
import static org.totschnig.myexpenses.provider.DatabaseConstants.KEY_ROWID;
import static org.totschnig.myexpenses.provider.DatabaseConstants.KEY_TEMPLATEID;

public class SplitPartList extends Fragment implements LoaderManager.LoaderCallbacks<Cursor> {
  private static final String KEY_PARENT_IS_TEMPLATE = "parentIsTemplate";
  private static final String KEY_ACCOUNT = "account";
  private static final int TRANSACTION_CURSOR = 5;
  private static final int SUM_CURSOR = 6;

  private SplitPartsListBinding binding;

  private SplitPartAdapter mAdapter;
  private long transactionSum = 0;
  private Money unsplitAmount;

  @State
  long parentId;

  @State
  long accountId;

  @Inject
  CurrencyFormatter currencyFormatter;

  public static @NonNull
  SplitPartList newInstance(Long transactionId, boolean isTemplate, Account account) {
    SplitPartList f = new SplitPartList();
    Bundle bundle = new Bundle();
    bundle.putLong(KEY_PARENTID, transactionId);
    bundle.putSerializable(KEY_ACCOUNT, account);
    bundle.putBoolean(KEY_PARENT_IS_TEMPLATE, isTemplate);
    f.setArguments(bundle);
    return f;
  }

  @Override
  public void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    setHasOptionsMenu(true);
    setRetainInstance(true);
    if (savedInstanceState == null) {
      parentId = requireArguments().getLong(KEY_PARENTID);
      accountId = requireArguments().getLong(KEY_ACCOUNTID);
    } else {
      Icepick.restoreInstanceState(this, savedInstanceState);
    }
    ((MyApplication) requireActivity().getApplication()).getAppComponent().inject(this);
  }

  @Override
  public void onSaveInstanceState(@NonNull Bundle outState) {
    super.onSaveInstanceState(outState);
    Icepick.saveInstanceState(this, outState);
  }

  @Override
  public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
    final ProtectedFragmentActivity ctx = (ProtectedFragmentActivity) getActivity();
    binding = SplitPartsListBinding.inflate(inflater, container, false);

    // Create an array to specify the fields we want to display in the list
    String[] from = new String[]{KEY_LABEL_MAIN, KEY_AMOUNT};
    // and an array of the fields we want to bind those fields to 
    int[] to = new int[]{R.id.category, R.id.amount};

    requireLoaders();
    // Now create a simple cursor adapter and set it to display
    final Account account = (Account) getArguments().getSerializable(KEY_ACCOUNT);
    mAdapter = new SplitPartAdapter(ctx, R.layout.split_part_row, null, from, to, 0,
        account.getCurrency(), currencyFormatter);
    binding.list.setAdapter(mAdapter);
    binding.list.setEmptyView(binding.empty);
    binding.list.setOnItemClickListener((a, v1, position, id) -> {
      Intent i = new Intent(ctx, ExpenseEdit.class);
      i.putExtra(parentIsTemplate() ? KEY_TEMPLATEID : KEY_ROWID, id);
      startActivityForResult(i, EDIT_REQUEST);
    });
    registerForContextMenu(binding.list);
    binding.CREATEPARTCOMMAND.setContentDescription(TextUtils.concatResStrings(getActivity(), ". ",
        R.string.menu_create_split_part_category, R.string.menu_create_split_part_transfer));
    return binding.getRoot();
  }

  @Override
  public void onDestroyView() {
    super.onDestroyView();
    binding = null;
  }

  @Override
  public void onCreateContextMenu(@NonNull ContextMenu menu, @NonNull View v,
                                  ContextMenuInfo menuInfo) {
    super.onCreateContextMenu(menu, v, menuInfo);
    menu.add(0, R.id.DELETE_COMMAND, 0, R.string.menu_delete);
  }

  @Override
  public boolean onContextItemSelected(android.view.MenuItem item) {
    AdapterContextMenuInfo info = (AdapterContextMenuInfo) item.getMenuInfo();
    if (item.getItemId() == R.id.DELETE_COMMAND) {
      ((ProtectedFragmentActivity) getActivity()).startTaskExecution(
          parentIsTemplate() ? TaskExecutionFragment.TASK_DELETE_TEMPLATES : TaskExecutionFragment.TASK_DELETE_TRANSACTION,
          new Long[]{info.id},
          Boolean.FALSE,
          0);
      return true;
    }
    return super.onContextItemSelected(item);
  }

  @NonNull
  @Override
  public Loader<Cursor> onCreateLoader(int id, Bundle args) {
    String[] selectionArgs = new String[]{String.valueOf(parentId)};
    CursorLoader cursorLoader = null;
    Uri uri = parentIsTemplate() ?
        TransactionProvider.TEMPLATES_UNCOMMITTED_URI : TransactionProvider.UNCOMMITTED_URI;
    switch (id) {
      case TRANSACTION_CURSOR:
        cursorLoader = new CursorLoader(getActivity(), uri, null, "parent_id = ?",
            selectionArgs, null);
        return cursorLoader;
      case SUM_CURSOR:
        cursorLoader = new CursorLoader(getActivity(), uri,
            new String[]{"sum(" + KEY_AMOUNT + ")"}, "parent_id = ?",
            selectionArgs, null);
    }
    return cursorLoader;
  }

  private boolean parentIsTemplate() {
    return getArguments().getBoolean(KEY_PARENT_IS_TEMPLATE);
  }

  @Override
  public void onLoadFinished(Loader<Cursor> arg0, Cursor c) {
    switch (arg0.getId()) {
      case TRANSACTION_CURSOR:
        mAdapter.swapCursor(c);
        break;
      case SUM_CURSOR:
        c.moveToFirst();
        transactionSum = c.getLong(0);
        updateBalance();
    }
  }

  @Override
  public void onLoaderReset(Loader<Cursor> arg0) {
    if (arg0.getId() == TRANSACTION_CURSOR) {
      mAdapter.swapCursor(null);
    }
  }

  public void updateBalance() {
    ExpenseEdit ctx = (ExpenseEdit) getActivity();
    if (ctx == null)
      return;
    unsplitAmount = ctx.getAmount();
    //when we are called before transaction is loaded in parent activity
    if (unsplitAmount == null)
      return;
    unsplitAmount = new Money(unsplitAmount.getCurrencyUnit(), unsplitAmount.getAmountMinor() - transactionSum);
    binding.end.setText(currencyFormatter.formatCurrency(unsplitAmount));
  }

  public boolean splitComplete() {
    return unsplitAmount != null && unsplitAmount.getAmountMinor() == 0L;
  }

  public int getSplitCount() {
    return mAdapter.getCount();
  }

  public void updateAccount(Account account) {
    accountId = account.getId();
    mAdapter.setCurrency(account.getCurrency());
    mAdapter.notifyDataSetChanged();
    updateBalance();
  }

  public void updateParent(long parentId) {
    this.parentId = parentId;
    requireLoaders();
  }

  private void requireLoaders() {
    Utils.requireLoader(LoaderManager.getInstance(getActivity()), TRANSACTION_CURSOR, null, this);
    Utils.requireLoader(LoaderManager.getInstance(getActivity()), SUM_CURSOR, null, this);
  }
}

File: myExpenses/src/main/java/org/totschnig/myexpenses/fragment/SplitPartList.kt
/*   This file is part of My Expenses.
 *   My Expenses is free software: you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License as published by
 *   the Free Software Foundation, either version 3 of the License, or
 *   (at your option) any later version.
 *
 *   My Expenses is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License
 *   along with My Expenses.  If not, see <http://www.gnu.org/licenses/>.
 */
package org.totschnig.myexpenses.fragment


import android.annotation.SuppressLint
import android.content.Intent
import android.os.Bundle
import android.view.ContextMenu
import android.view.ContextMenu.ContextMenuInfo
import android.view.LayoutInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import icepick.Icepick
import icepick.State
import org.totschnig.myexpenses.MyApplication
import org.totschnig.myexpenses.R
import org.totschnig.myexpenses.activity.EDIT_REQUEST
import org.totschnig.myexpenses.activity.EditActivity
import org.totschnig.myexpenses.activity.ExpenseEdit
import org.totschnig.myexpenses.activity.ProtectedFragmentActivity
import org.totschnig.myexpenses.adapter.SplitPartRVAdapter
import org.totschnig.myexpenses.databinding.SplitPartsListBinding
import org.totschnig.myexpenses.model.CurrencyUnit
import org.totschnig.myexpenses.model.Money
import org.totschnig.myexpenses.provider.DatabaseConstants
import org.totschnig.myexpenses.provider.DatabaseConstants.KEY_CURRENCY
import org.totschnig.myexpenses.provider.DatabaseConstants.KEY_ROWID
import org.totschnig.myexpenses.provider.DatabaseConstants.KEY_TEMPLATEID
import org.totschnig.myexpenses.ui.ContextAwareRecyclerView
import org.totschnig.myexpenses.util.CurrencyFormatter
import org.totschnig.myexpenses.util.TextUtils.concatResStrings
import org.totschnig.myexpenses.util.formatMoney
import org.totschnig.myexpenses.viewmodel.SplitPartListViewModel
import org.totschnig.myexpenses.viewmodel.data.Account
import javax.inject.Inject

class SplitPartList : Fragment() {
    private var _binding: SplitPartsListBinding? = null
    private val binding get() = _binding!!
    lateinit var viewModel: SplitPartListViewModel
    lateinit var adapter: SplitPartRVAdapter
    private var transactionSum: Long = 0
    private var unsplitAmount: Money? = null

    @JvmField
    @State
    var parentId: Long = 0

    @Inject
    lateinit var currencyFormatter: CurrencyFormatter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setHasOptionsMenu(true)
        if (savedInstanceState == null) {
            parentId = requireArguments().getLong(DatabaseConstants.KEY_PARENTID)
        } else {
            Icepick.restoreInstanceState(this, savedInstanceState)
        }
        val appComponent = (requireActivity().application as MyApplication).appComponent
        appComponent.inject(this)
        viewModel = ViewModelProvider(this)[SplitPartListViewModel::class.java]
        appComponent.inject(viewModel)
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        Icepick.saveInstanceState(this, outState)
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        val ctx = requireActivity() as ProtectedFragmentActivity
        _binding = SplitPartsListBinding.inflate(inflater, container, false)
        val currencyUnit = requireArguments().getSerializable(KEY_CURRENCY) as CurrencyUnit
        adapter = SplitPartRVAdapter(
            requireContext(),
            currencyUnit,
            currencyFormatter
        ) { view, _ -> requireActivity().openContextMenu(view) }
        binding.list.adapter = adapter
        viewModel.getSplitParts().observe(
            viewLifecycleOwner, { transactions ->
                binding.empty.visibility = if (transactions.isEmpty()) View.VISIBLE else View.GONE
                binding.list.visibility = if (transactions.isEmpty()) View.GONE else View.VISIBLE
                transactionSum = transactions.sumOf { it.amountRaw }
                adapter.submitList(transactions)
                updateBalance()
            })
        loadParts()
        registerForContextMenu(binding.list)
        binding.CREATEPARTCOMMAND.contentDescription = concatResStrings(
            ctx, ". ",
            R.string.menu_create_split_part_category, R.string.menu_create_split_part_transfer
        )
        return binding.root
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }

    override fun onCreateContextMenu(
        menu: ContextMenu, v: View,
        menuInfo: ContextMenuInfo?
    ) {
        super.onCreateContextMenu(menu, v, menuInfo)
        menu.add(0, R.id.EDIT_COMMAND, 0, R.string.menu_edit)
        menu.add(0, R.id.DELETE_COMMAND, 0, R.string.menu_delete)
    }

    override fun onContextItemSelected(item: MenuItem): Boolean {
        val info = item.menuInfo as ContextAwareRecyclerView.RecyclerContextMenuInfo
        return when (item.itemId) {
            R.id.EDIT_COMMAND -> {
                startActivityForResult(Intent(requireContext(), ExpenseEdit::class.java).apply {
                    putExtra(if (parentIsTemplate()) KEY_TEMPLATEID else KEY_ROWID, info.id)
                }, EDIT_REQUEST)
                true
            }
            R.id.DELETE_COMMAND -> {
                val resultObserver = Observer { result: Int ->
                    val activity = requireActivity() as EditActivity
                    if (result > 0) {
                        activity.showSnackbar(
                            activity.resources.getQuantityString(
                                R.plurals.delete_success,
                                result,
                                result
                            )
                        )
                        activity.setDirty()
                    } else {
                        activity.showDeleteFailureFeedback(null)
                    }
                }
                if (parentIsTemplate()) {
                    viewModel.deleteTemplates(longArrayOf(info.id), false)
                        .observe(viewLifecycleOwner, resultObserver)
                } else {
                    viewModel.deleteTransactions(longArrayOf(info.id), false).observe(
                        viewLifecycleOwner, resultObserver
                    )
                }
                true
            }
            else -> super.onContextItemSelected(item)
        }
    }

    private fun parentIsTemplate(): Boolean {
        return requireArguments().getBoolean(KEY_PARENT_IS_TEMPLATE)
    }

    fun updateBalance() {
        (activity as? ExpenseEdit)?.amount?.let {
            unsplitAmount = Money(it.currencyUnit, it.amountMinor - transactionSum)
            binding.end.text = unsplitAmountFormatted
        }
    }

    val unsplitAmountFormatted
        get() = unsplitAmount?.let { currencyFormatter.formatMoney(it) }

    val splitComplete
        get() = unsplitAmount?.amountMinor == 0L

    val splitCount: Int
        get() = adapter.itemCount

    @SuppressLint("NotifyDataSetChanged")
    fun updateAccount(account: Account) {
        adapter.currencyUnit = account.currency
        adapter.notifyDataSetChanged()
        updateBalance()
    }

    fun updateParent(parentId: Long) {
        this.parentId = parentId
        loadParts()
    }

    private fun loadParts() {
        viewModel.loadSplitParts(parentId, parentIsTemplate())
    }

    companion object {
        private const val KEY_PARENT_IS_TEMPLATE = "parentIsTemplate"
        fun newInstance(
            transactionId: Long,
            isTemplate: Boolean,
            currencyUnit: CurrencyUnit
        ) = SplitPartList().apply {
            arguments = Bundle().apply {
                putLong(DatabaseConstants.KEY_PARENTID, transactionId)
                putSerializable(KEY_CURRENCY, currencyUnit)
                putBoolean(KEY_PARENT_IS_TEMPLATE, isTemplate)
            }
        }
    }
}